File: typescript/packages/extensions/src/sign-in-with-x/hooks.ts
/**
 * SIWX Lifecycle Hooks
 *
 * Pre-built hooks for integrating SIWX authentication with x402 servers and clients.
 */

import type { SIWxStorage } from "./storage";
import type { SIWxExtension, SIWxVerifyOptions, SignatureType } from "./types";
import type { SIWxSigner } from "./sign";
import type { ClientExtension } from "@x402/core/client";
import type { PaymentRequiredContext } from "@x402/core/http";
import type { ProtectedRequestHook } from "@x402/core/http";
import { SIGN_IN_WITH_X } from "./types";
import { parseSIWxHeader } from "./parse";
import { validateSIWxMessage } from "./validate";
import { verifySIWxSignature } from "./verify";
import { createSIWxPayload } from "./client";
import { encodeSIWxHeader } from "./encode";
import { isSolanaSigner } from "./solana";

/**
 * Normalizes and validates a configured SIWX origin.
 *
 * @param origin - Absolute http(s) origin without credentials, path, query, or fragment
 * @returns Parsed origin URL
 * @throws Error when the origin string is invalid
 */
export function normalizeConfiguredOrigin(origin: string): URL {
  let url: URL;
  try {
    url = new URL(origin);
  } catch {
    throw new Error(`Invalid SIWX origin: "${origin}" is not a valid URL`);
  }

  if (url.protocol !== "http:" && url.protocol !== "https:") {
    throw new Error(`Invalid SIWX origin: "${origin}" must use http: or https:`);
  }

  if (url.username || url.password) {
    throw new Error(`Invalid SIWX origin: "${origin}" must not include credentials`);
  }

  if (url.pathname !== "/" || url.search || url.hash) {
    throw new Error(`Invalid SIWX origin: "${origin}" must not include a path, query, or fragment`);
  }

  return url;
}

/**
 * Options for creating the SIWX settle hook.
 */
export interface CreateSIWxSettleHookOptions {
  /** Storage for tracking paid addresses */
  storage: SIWxStorage;
  /** Optional callback for logging/debugging */
  onEvent?: (event: SIWxHookEvent) => void;
}

/**
 * Options for creating the SIWX request hook and resource server extension.
 */
export interface CreateSIWxRequestHookOptions extends CreateSIWxSettleHookOptions {
  /**
   * Public browser-visible origin for SIWX domain and URI binding.
   *
   * Set this to the external origin (for example `https://api.example.com`),
   * not the upstream listener address behind a reverse proxy.
   */
  origin: string;
  /** Options for signature verification (e.g., EVM smart wallet support) */
  verifyOptions?: SIWxVerifyOptions;
}

export type CreateSIWxHookOptions = CreateSIWxRequestHookOptions;

/**
 * Options for creating the SIWX client extension.
 */
export interface CreateSIWxClientExtensionOptions {
  /** Wallet signers to try against the server's supported SIWX chains */
  signers: SIWxSigner[];
}

/**
 * Events emitted by SIWX hooks for logging/debugging.
 */
export type SIWxHookEvent =
  | { type: "payment_recorded"; resource: string; address: string }
  | { type: "access_granted"; resource: string; address: string }
  | { type: "validation_failed"; resource: string; error?: string }
  | { type: "nonce_reused"; resource: string; nonce: string }
  | { type: "siwx_header_sent"; resource: string };

/**
 * Creates an onAfterSettle hook that records payments for SIWX.
 *
 * @param options - Hook configuration
 * @returns Hook function for x402ResourceServer.onAfterSettle()
 *
 * @example
 * ```typescript
 * const storage = new InMemorySIWxStorage();
 * const resourceServer = new x402ResourceServer(facilitator)
 *   .onAfterSettle(createSIWxSettleHook({ storage }));
 * ```
 */
export function createSIWxSettleHook(options: CreateSIWxSettleHookOptions) {
  const { storage, onEvent } = options;

  return async (ctx: {
    paymentPayload: { payload: unknown; resource?: { url: string } };
    result: { success: boolean; payer?: string };
  }): Promise<void> => {
    // Only record payment if settlement succeeded
    if (!ctx.result.success) return;

    // Get payer from facilitator's settle result (works for all payment schemes)
    const address = ctx.result.payer;
    if (!address) return;

    // resource is optional per the v2 spec (section 5.2.2)
    const resourceUrl = ctx.paymentPayload.resource?.url;
    if (!resourceUrl) return;

    const resource = new URL(resourceUrl).pathname;
    await storage.recordPayment(resource, address);
    onEvent?.({ type: "payment_recorded", resource, address });
  };
}

/**
 * Creates an onProtectedRequest hook that validates SIWX auth.
 *
 * For paid routes: grants access when the SIWX signature is valid and the address has paid.
 * For auth-only routes (accepts: []): grants access on valid SIWX signature alone.
 * Auth-only detection uses the routeConfig passed by x402HTTPResourceServer.
 *
 * @param options - Hook configuration including required configured origin
 * @returns Hook function for x402HTTPResourceServer.onProtectedRequest()
 *
 * @example
 * ```typescript
 * const storage = new InMemorySIWxStorage();
 * const httpServer = new x402HTTPResourceServer(resourceServer, routes)
 *   .onProtectedRequest(createSIWxRequestHook({
 *     storage,
 *     origin: "https://api.example.com",
 *   }));
 * ```
 */
export function createSIWxRequestHook(options: CreateSIWxRequestHookOptions): ProtectedRequestHook {
  const { storage, verifyOptions, onEvent } = options;
  const configuredOrigin = normalizeConfiguredOrigin(options.origin);

  // Validate nonce tracking is fully implemented or not at all
  const hasUsedNonce = typeof storage.hasUsedNonce === "function";
  const hasRecordNonce = typeof storage.recordNonce === "function";
  if (hasUsedNonce !== hasRecordNonce) {
    throw new Error(
      "SIWxStorage nonce tracking requires both hasUsedNonce and recordNonce to be implemented",
    );
  }

  return async (context, routeConfig) => {
    // Try both cases for header (HTTP headers are case-insensitive)
    const header =
      context.adapter.getHeader(SIGN_IN_WITH_X) ||
      context.adapter.getHeader(SIGN_IN_WITH_X.toLowerCase());
    if (!header) return;

    try {
      const payload = parseSIWxHeader(header);
      const validation = await validateSIWxMessage(payload, configuredOrigin);
      if (!validation.valid) {
        onEvent?.({ type: "validation_failed", resource: context.path, error: validation.error });
        return;
      }

      const verification = await verifySIWxSignature(payload, verifyOptions);
      if (!verification.valid || !verification.address) {
        onEvent?.({ type: "validation_failed", resource: context.path, error: verification.error });
        return;
      }

      // Check if nonce was already used (prevents signature replay attacks)
      if (storage.hasUsedNonce) {
        const nonceUsed = await storage.hasUsedNonce(payload.nonce);
        if (nonceUsed) {
          onEvent?.({ type: "nonce_reused", resource: context.path, nonce: payload.nonce });
          return;
        }
      }

      // Auth-only routes (accepts: []) grant access on valid SIWX alone
      const isAuthOnly = Array.isArray(routeConfig.accepts) && routeConfig.accepts.length === 0;

      const shouldGrant = isAuthOnly || (await storage.hasPaid(context.path, verification.address));
      if (shouldGrant) {
        // Record nonce as used before granting access
        if (storage.recordNonce) {
          await storage.recordNonce(payload.nonce);
        }

        onEvent?.({
          type: "access_granted",
          resource: context.path,
          address: verification.address,
        });
        return { grantAccess: true };
      }
    } catch (err) {
      onEvent?.({
        type: "validation_failed",
        resource: context.path,
        error: err instanceof Error ? err.message : "Unknown error",
      });
    }
  };
}

/**
 * Creates an onPaymentRequired hook for client-side SIWX authentication.
 *
 * Matches the signer type to a compatible chain in supportedChains.
 * For EVM signers: matches any eip191 chain
 * For Solana signers: matches any ed25519 chain
 *
 * @param signer - Wallet signer for creating SIWX proofs
 * @returns Hook function for x402HTTPClient.onPaymentRequired()
 *
 * @example
 * ```typescript
 * const httpClient = new x402HTTPClient(client)
 *   .onPaymentRequired(createSIWxClientHook(signer));
 * ```
 */
export function createSIWxClientHook(signer: SIWxSigner) {
  // Determine signer type once at hook creation
  const signerIsSolana = isSolanaSigner(signer);
  const expectedSignatureType: SignatureType = signerIsSolana ? "ed25519" : "eip191";

  return async (context: {
    paymentRequired: { accepts?: Array<{ network: string }>; extensions?: Record<string, unknown> };
  }): Promise<{ headers: Record<string, string> } | void> => {
    const extensions = context.paymentRequired.extensions ?? {};
    const siwxExtension = extensions[SIGN_IN_WITH_X] as SIWxExtension | undefined;

    if (!siwxExtension?.supportedChains) return;

    try {
      // Find a chain that matches the signer's signature type
      const matchingChain = siwxExtension.supportedChains.find(
        chain => chain.type === expectedSignatureType,
      );

      if (!matchingChain) {
        // No chain compatible with this signer type
        return;
      }

      // Build complete info with selected chain
      const completeInfo = {
        ...siwxExtension.info,
        chainId: matchingChain.chainId,
        type: matchingChain.type,
      };

      const payload = await createSIWxPayload(completeInfo, signer);
      const header = encodeSIWxHeader(payload);
      return { headers: { [SIGN_IN_WITH_X]: header } };
    } catch {
      // Failed to create SIWX payload, continue to payment
    }
  };
}

/**
 * Creates a SIWX client extension that signs HTTP SIWX challenges for compatible wallets.
 *
 * @param options - Client extension configuration (signers tried in order until one succeeds)
 * @returns x402 client extension registering HTTP transport hooks for SIWX
 */
export function createSIWxClientExtension(
  options: CreateSIWxClientExtensionOptions,
): ClientExtension {
  const hooks = options.signers.map(createSIWxClientHook);

  return {
    key: SIGN_IN_WITH_X,
    transportHooks: {
      http: {
        onPaymentRequired: async (_declaration: unknown, context: PaymentRequiredContext) => {
          for (const hook of hooks) {
            const result = await hook(context);
            if (result?.headers) return result;
          }
        },
      },
    },
  };
}

File: python/x402/extensions/sign_in_with_x/hooks.py
"""SIWX lifecycle hooks."""

from __future__ import annotations

import inspect
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any
from urllib.parse import urlparse

from x402.http.types import HTTPRequestContext, PaymentOption, RouteConfig
from x402.schemas.extensions import ClientExtension
from x402.schemas.hooks import GrantAccessResult, PaymentRequiredHeadersResult

from .client import create_siwx_payload
from .encode import encode_siwx_header
from .parse import parse_siwx_header
from .solana import is_solana_signer
from .types import SIGN_IN_WITH_X, SignatureType, SIWxVerifyOptions
from .validate import validate_siwx_message
from .verify import verify_siwx_signature

SIWxHookEvent = dict[str, Any]


def normalize_configured_origin(origin: str) -> str:
    """Normalize and validate a configured SIWX origin."""
    parsed = urlparse(origin)

    if parsed.scheme not in ("http", "https") or not parsed.netloc:
        if parsed.scheme and parsed.scheme not in ("http", "https"):
            raise ValueError(f'Invalid SIWX origin: "{origin}" must use http: or https:')
        raise ValueError(f'Invalid SIWX origin: "{origin}" is not a valid URL')

    if parsed.username or parsed.password:
        raise ValueError(f'Invalid SIWX origin: "{origin}" must not include credentials')

    if parsed.path not in ("", "/") or parsed.query or parsed.fragment:
        raise ValueError(
            f'Invalid SIWX origin: "{origin}" must not include a path, query, or fragment'
        )

    return f"{parsed.scheme}://{parsed.netloc}"


@dataclass(kw_only=True)
class CreateSIWxSettleHookOptions:
    """Options for creating the SIWX settle hook."""

    storage: Any
    on_event: Callable[[SIWxHookEvent], None] | None = None


@dataclass(kw_only=True)
class CreateSIWxRequestHookOptions(CreateSIWxSettleHookOptions):
    """Options for creating the SIWX request hook and resource server extension."""

    origin: str
    verify_options: SIWxVerifyOptions | None = None


CreateSIWxHookOptions = CreateSIWxRequestHookOptions


@dataclass
class CreateSIWxClientExtensionOptions:
    """Options for creating the SIWX client extension."""

    signers: list[Any]


def create_siwx_settle_hook(options: CreateSIWxSettleHookOptions):
    """Create an onAfterSettle hook that records payments for SIWX."""

    async def hook(ctx: Any) -> None:
        if not ctx.result.success:
            return
        address = ctx.result.payer
        if not address:
            return
        resource = getattr(ctx.payment_payload, "resource", None)
        resource_url = resource.url if resource else None
        if not resource_url:
            return
        path = urlparse(resource_url).path
        storage = options.storage
        result = storage.record_payment(path, address)
        if inspect.isawaitable(result):
            await result
        if options.on_event:
            options.on_event({"type": "payment_recorded", "resource": path, "address": address})

    return hook


def create_siwx_request_hook(options: CreateSIWxRequestHookOptions):
    """Create an onProtectedRequest hook that validates SIWX auth."""
    configured_origin = normalize_configured_origin(options.origin)
    storage = options.storage
    has_used_nonce = callable(getattr(storage, "has_used_nonce", None))
    has_record_nonce = callable(getattr(storage, "has_record_nonce", None))
    if has_used_nonce != has_record_nonce:
        raise ValueError(
            "SIWxStorage nonce tracking requires both has_used_nonce and record_nonce "
            "to be implemented"
        )

    async def hook(
        context: HTTPRequestContext,
        route_config: RouteConfig | None = None,
    ) -> GrantAccessResult | None:
        adapter = context.adapter
        header = adapter.get_header(SIGN_IN_WITH_X) or adapter.get_header(SIGN_IN_WITH_X.lower())
        if not header:
            return None

        try:
            payload = parse_siwx_header(header)
            validation = await validate_siwx_message(payload, configured_origin)
            if not validation.valid:
                if options.on_event:
                    options.on_event(
                        {
                            "type": "validation_failed",
                            "resource": context.path,
                            "error": validation.error,
                        }
                    )
                return None

            verification = await verify_siwx_signature(payload, options.verify_options)
            if not verification.valid or not verification.address:
                if options.on_event:
                    options.on_event(
                        {
                            "type": "validation_failed",
                            "resource": context.path,
                            "error": verification.error,
                        }
                    )
                return None

            if has_used_nonce:
                nonce_used = storage.has_used_nonce(payload.nonce)
                if inspect.isawaitable(nonce_used):
                    nonce_used = await nonce_used
                if nonce_used:
                    if options.on_event:
                        options.on_event(
                            {
                                "type": "nonce_reused",
                                "resource": context.path,
                                "nonce": payload.nonce,
                            }
                        )
                    return None

            accepts = route_config.accepts if route_config else None
            if isinstance(accepts, PaymentOption):
                accept_list: list[Any] = [accepts]
            else:
                accept_list = list(accepts or [])
            is_auth_only = isinstance(accept_list, list) and len(accept_list) == 0

            has_paid = storage.has_paid(context.path, verification.address)
            if inspect.isawaitable(has_paid):
                has_paid = await has_paid
            should_grant = is_auth_only or has_paid
            if should_grant:
                if has_record_nonce:
                    record = storage.record_nonce(payload.nonce)
                    if inspect.isawaitable(record):
                        await record
                if options.on_event:
                    options.on_event(
                        {
                            "type": "access_granted",
                            "resource": context.path,
                            "address": verification.address,
                        }
                    )
                return GrantAccessResult()
        except Exception as err:
            if options.on_event:
                options.on_event(
                    {
                        "type": "validation_failed",
                        "resource": context.path,
                        "error": str(err),
                    }
                )
        return None

    return hook


def create_siwx_client_hook(signer: Any):
    """Create an onPaymentRequired hook for client-side SIWX authentication."""
    signer_is_solana = is_solana_signer(signer)
    expected_signature_type: SignatureType = "ed25519" if signer_is_solana else "eip191"

    async def hook(context: Any) -> PaymentRequiredHeadersResult | None:
        extensions = context.payment_required.extensions or {}
        siwx_extension = extensions.get(SIGN_IN_WITH_X)
        if not siwx_extension:
            return None
        if isinstance(siwx_extension, dict):
            supported = siwx_extension.get("supportedChains") or []
            info = siwx_extension.get("info") or {}
        else:
            supported = siwx_extension.supported_chains
            info = siwx_extension.info.model_dump(by_alias=True, exclude_none=True)

        if not supported:
            return None

        try:
            matching = next(
                (
                    c
                    for c in supported
                    if (c.get("type") if isinstance(c, dict) else c.type) == expected_signature_type
                ),
                None,
            )
            if matching is None:
                return None

            chain_id = matching["chainId"] if isinstance(matching, dict) else matching.chain_id
            sig_type = matching["type"] if isinstance(matching, dict) else matching.type
            complete_info = {**info, "chainId": chain_id, "type": sig_type}
            payload = await create_siwx_payload(complete_info, signer)
            header = encode_siwx_header(payload)
            return PaymentRequiredHeadersResult(headers={SIGN_IN_WITH_X: header})
        except Exception:
            return None

    return hook


def create_siwx_client_extension(options: CreateSIWxClientExtensionOptions) -> ClientExtension:
    """Create a SIWX client extension with HTTP transport hooks."""
    hooks = [create_siwx_client_hook(signer) for signer in options.signers]

    class _SIWxClientExtension:
        key = SIGN_IN_WITH_X

        @property
        def transport_hooks(self) -> _SIWxClientTransportHooks:
            return _SIWxClientTransportHooks(hooks=hooks)

    return _SIWxClientExtension()


@dataclass
class _SIWxClientTransportHooks:
    hooks: list[Any]

    @property
    def http(self) -> _SIWxClientHTTPHooks:
        return _SIWxClientHTTPHooks(self.hooks)


@dataclass
class _SIWxClientHTTPHooks:
    hooks: list[Any]

    async def on_payment_required(
        self, _declaration: Any, context: Any
    ) -> PaymentRequiredHeadersResult | None:
        for hook in self.hooks:
            result = await hook(context)
            if result is not None:
                return result
        return None
